src/gluonnlp/models/t5.py

```diff
@@ -0,0 +1,883 @@
+def google_t5_large(): 
+    cfg = google_t5_base()
+    cfg.defrost()
+    cfg.MODEL.d_model = 1024
+    cfg.MODEL.d_ff = 4096
+    cfg.MODEL.num_layers = 24
+    cfg.MODEL.num_heads = 16
+    cfg.freeze()
+    return cfg
+
+
+@t5_cfg_reg.register()
+def google_t5_3B(): 
+    cfg = google_t5_base()
+    cfg.defrost()
+    cfg.MODEL.d_model = 1024
+    cfg.MODEL.d_kv = 128
+    cfg.MODEL.d_ff = 16384
+    cfg.MODEL.num_layers = 24
+    cfg.MODEL.num_heads = 32
+    cfg.freeze()
+    return cfg
+
+
+@t5_cfg_reg.register()
+def google_t5_11B(): 
+    cfg = google_t5_base()
+    cfg.defrost()
+    cfg.MODEL.d_model = 1024
+    cfg.MODEL.d_kv = 128
+    cfg.MODEL.d_ff = 65536
+    cfg.MODEL.num_layers = 24
+    cfg.MODEL.num_heads = 128
+    cfg.freeze()
+    return cfg
+
+
+PRETRAINED_URL = {
+    'google_t5_small': {
+        'cfg': google_t5_small(), 
+        'vocab': 'google_t5_small/t5-5f05e7c5.vocab', 
+        'params': 'google_t5_small/model-e34f6fbd.params'
+    }, 
+    'google_t5_base': {
+        'cfg': google_t5_base(), 
+        'vocab': 'google_t5_base/t5-5f05e7c5.vocab', 
+        'params': 'google_t5_base/model-e1956ac9.params'
+    }, 
+    'google_t5_large': {
+        'cfg': google_t5_large(), 
+        'vocab': 'google_t5_large/t5-5f05e7c5.vocab', 
+        'params': 'google_t5_large/model-bf5fc813.params'
+    }, 
+    'google_t5_3B': {
+        'cfg': google_t5_3B()
+    }, 
+    'google_t5_11B': {
+        'cfg': google_t5_11B()
+    }
+}
+
+
+@use_np
+class T5Block(HybridBlock): 
+    def __init__(
+        self, 
+        d_model, 
+        d_kv, 
+        d_ff, 
+        is_decoder, 
+        num_heads=12, 
+        dropout_prob=0.1, 
+        layer_norm_eps=1E-6, 
+        activation='relu', 
+        init_factor=1.0, 
+        layout='NT', 
+        dtype='float32'
+    ): 
+        super().__init__()
+        self._d_model = d_model
+        self._d_kv = d_kv
+        self._d_ff = d_ff
+        self._is_decoder = is_decoder
+        self._num_heads = num_heads
+        self._inner_dim = self._num_heads * self._d_kv
+        self._dtype = dtype
+        assert layout in ['TN', 'NT'], \
+            'Invalid layout: {}. Only "TN" and "NT" are supported.'.format(layout)
+        self._layout = layout
+
+        self.self_attn_layer_norm = RMSNorm(
+            in_channels=d_model, 
+            center=False, 
+            scale=True, 
+            gamma_initializer=Constant(1.0 * init_factor), 
+            variance_epsilon=layer_norm_eps, 
+            dtype=dtype
+        )
+        # avoid scaling before softmax
+        # See https://github.com/tensorflow/mesh/blob/fa19d69eafc9a482aff0b59ddd96b025c0cb207d/mesh_tensorflow/transformer/attention.py#L136
+        self.self_attn_q = nn.Dense(
+                units=self._inner_dim, 
+                in_units=d_model, 
+                flatten=False, 
+                use_bias=False, 
+                weight_initializer=Normal((d_model * d_kv) ** -0.5 * init_factor), 
```

Comment:
@ZheyuYe @leezu @szhengac @szha Just mention here in case you do not konw. T5 used this trick to avoid dividing the `sqrt(query_dim)` as in `softmax(<query, key> / sqrt(query_dim))`. Basically, the trick is to move `1/sqrt(query_dim)` to the initializer.

Reply:
They don't seem equivalent

Reply:
I leave it consistent with huggingface's implementation.

Reply:
Since all query are generated as `W_q X`. Calculating `W_q X / sqrt(query_dim)` is equivalent to `W_q / sqrt(query_dim) X`. We can then move `1/sqrt(query_dim)` to the initialization phase of `W_q` so that they will be the same when we first initialize the model.

Reply:
@szha Would you give more detail about the difference between these two approaches? I feel that they give us the same initialization of the attention scores.

Reply:
The obvious difference is that the gradients would be of different scales.

Reply:
Have you seen any difference other than the scale of the gradient? For me, I feel that the main point is that this ensures that the softmax attention weights are initialized similarly.

Reply:
I haven't tried this myself and the following is just based on reasoning. Since the gradients are different, I'd expect this parameter to be trained differently beyond initialization. In your example, the new gradient of W_q after change would be `X` instead of `X / sqrt(query_dim)`. Then, behaviors such as clipping global gradient norm and regularization will change too.